californium-core/src/main/java/org/eclipse/californium/core/network/stack/BlockwiseLayer.java

```diff
@@ -527,302 +620,328 @@ private void handleBlock2Response(final Exchange exchange, final Response respon
+					int currentSize = status.getCurrentSize();
+					// do late block size negotiation
+					int newSize, newSzx;
+					if (block2.getSzx() > preferredBlockSzx) {
+						newSize = preferredBlockSize;
+						newSzx = preferredBlockSzx;
+					} else {
+						newSize = currentSize;
+						newSzx = status.getCurrentSzx();
+					}
+					int nextNum = status.getCurrentNum() + currentSize / newSize;
+
+					Request request = exchange.getRequest();
+
+					Request block = new Request(request.getCode());
+					// do not enforce CON, since NON could make sense over SMS or similar transports
+					block.setType(request.getType());
+					block.setDestination(request.getDestination());
+					block.setDestinationPort(request.getDestinationPort());
+
+					/*
+					 * WARNING:
+					 * 
+					 * For Observe, the Matcher then will store the same
+					 * exchange under a different KeyToken in exchangesByToken,
+					 * which is cleaned up in the else case below.
+					 */
+					if (!response.getOptions().hasObserve()) {
+						block.setToken(response.getToken());
+					}
 
-				LOGGER.log(Level.FINER, "Requesting next Block2 num={0}", num);
+					// copy options
+					block.setOptions(new OptionSet(request.getOptions()));
+					block.getOptions().setBlock2(newSzx, false, nextNum);
+					if (response.getOptions().getETagCount() > 0) {
+						// use ETag provided by peer
+						block.getOptions().addETag(response.getOptions().getETags().get(0));
+					}
 
-				Request block = new Request(request.getCode());
-				// do not enforce CON, since NON could make sense over SMS or similar transports
-				block.setType(request.getType());
-				block.setDestination(request.getDestination());
-				block.setDestinationPort(request.getDestinationPort());
+					// make sure NOT to use Observe for block retrieval
+					block.getOptions().removeObserve();
 
-				/*
-				 * WARNING:
-				 * 
-				 * For Observe, the Matcher then will store the same
-				 * exchange under a different KeyToken in exchangesByToken,
-				 * which is cleaned up in the else case below.
-				 */
-				if (!response.getOptions().hasObserve()) block.setToken(response.getToken());
+					// copy message observers from original request so that they will be notified
+					// if something goes wrong with this blockwise request, e.g. if it times out
+					block.addMessageObservers(request.getMessageObservers());
+					// add an observer that cleans up the block2 transfer tracker if the
+					// block request fails
+					addBlock2CleanUpObserver(block, key);
 
-				// copy options
-				block.setOptions(new OptionSet(request.getOptions()));
-				// make sure NOT to use Observe for block retrieval
-				block.getOptions().removeObserve();
+					status.setCurrentNum(nextNum);
 
-				block.getOptions().setBlock2(szx, m, num);
+					LOGGER.log(Level.FINER, "requesting next Block2 [num={0}]: {1}", new Object[]{ nextNum, block });
+					exchange.setCurrentRequest(block);
+					lower().sendRequest(exchange, block);
```

Comment:
This is done in the `handleBlock2Response()` method. More specifically, in the branch where the last block is handled:
```

				} else {

					// we have received the last block of the block2 transfer

					LOGGER.log(
							Level.FINER,
							"all {0} blocks have been retrieved, assembling response and delivering to application layer",
							status.getBlockCount());
					Response assembled = new Response(response.getCode());
					status.assembleMessage(assembled);

					// set overall transfer RTT
					assembled.setRTT(System.currentTimeMillis() - exchange.getTimestamp());

					if (status.isNotification()) {

						/*
						 * When retrieving the rest of a blockwise notification
						 * with a different token, the additional Matcher state
						 * must be cleaned up through the call below.
						 */
						if (!response.getOptions().hasObserve()) {
							// call the clean-up mechanism for the additional Matcher entry in exchangesByToken
							exchange.completeCurrentRequest();
						}
					}

					clearBlock2Status(key);
					LOGGER.log(Level.FINE, "assembled response: {0}", assembled);
					// Set the original request as current request so that
					// the Matcher can clean up its state based on the latest
					// ("current") request's MID and token
					exchange.setCurrentRequest(exchange.getRequest());
					// Set the assembled response as current response
					exchange.setResponse(assembled);
					upper().receiveResponse(exchange, assembled);
				}

```
Since we are using the same token for retrieving all blocks it is sufficient to clean up once the last block has been received.